File: 프로그래머스/lv2_위장.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>

using namespace std;

int solution(vector<vector<string>> clothes) {
    int answer = 0;
    unordered_map<string, int> hashmap;

    // 옷 종류의 개수 계산
    for( int i=0; i<clothes.size(); i++ ){
        hashmap[clothes[i][1]]++;
    }

    answer = 1;
    for( auto elem : hashmap ){
        answer *= elem.second + 1; // +1 은 선택하지 않음에 해당하는 경우의 수
    }
    answer -= 1; // -1 은 모든 경우에서 선택하지 않음. 즉, 아무것도 입지 않음에 대한 경우의 수
    
    return answer;
}

int main(void){

    vector<vector<string>> clothes = {{"yellowhat", "headgear"}, {"bluesunglasses", "eyewear"}, {"green_turban", "headgear"}};
    int ret;

    ret = solution(clothes);
    cout << ret << endl;

    return 0;
}

File: 프로그래머스/lv2_위장.py
from functools import reduce
from collections import Counter
from itertools import combinations

'''
    조합의 갯수만 찾으면 되기 때문에, 경우의 수 공식 이용    
'''
def solution(clothes):
    answer = 0

    cnt = Counter([kind for name, kind in clothes])

    answer = reduce(lambda x,y : x*(y+1), cnt.values(), 1) - 1

    return answer 


'''
# 가능한 모든 조합을 구해서 갯수를 세는 것은 시간초과 발생
def solution(clothes):
    answer = 0

    dic = {}
    for x,y in clothes:
        dic[x] = y

    answer += len(clothes)
    for i in range(2,len(dic.keys())+1):
        for c in combinations(dic.values(),i):
            check = {}
            for elem in c:
                if (elem not in check):
                    check[elem] = 1
                else:
                    break
            else:
                print(c)
                answer += 1

    return answer
'''

if __name__ == "__main__":
    clothes = [["yellowhat", "headgear"], ["bluesunglasses", "eyewear"], ["green_turban", "headgear"]]
    ret = solution(clothes)

    print( ret )